x/wasm/keeper/msg_server.go

```diff
@@ -74,6 +75,37 @@ func (m msgServer) InstantiateContract(goCtx context.Context, msg *types.MsgInst
 	}, nil
 }
 
+// InstantiateContract2 instantiate a new contract with predicatable address generated
+func (m msgServer) InstantiateContract2(goCtx context.Context, msg *types.MsgInstantiateContract2) (*types.MsgInstantiateContract2Response, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	senderAddr, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "sender")
+	}
+	var adminAddr sdk.AccAddress
+	if msg.Admin != "" {
+		if adminAddr, err = sdk.AccAddressFromBech32(msg.Admin); err != nil {
+			return nil, sdkerrors.Wrap(err, "admin")
+		}
+	}
+
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
```

Comment:
This event is combined with the message type, right?

We need to document that people basing on events will have to search for two event values now.

Reply:
ja, see below (this is the list of events for the `hackatom` example)

```json
[
  {
    "events": [
      {
        "type": "coin_received",
        "attributes": [
          {
            "key": "receiver",
            "value": "wasm1ayj3kc8grpe7axlvdhc0etrhev9knjs0v4ghwc0qn8hda3mdathq75ahs3"
          },
          {
            "key": "amount",
            "value": "100ustake"
          }
        ]
      },
      {
        "type": "coin_spent",
        "attributes": [
          {
            "key": "spender",
            "value": "wasm1alz4eapgf0rq3ezf3xge9pz7uhkpdyu2mmrrup"
          },
          {
            "key": "amount",
            "value": "100ustake"
          }
        ]
      },
      {
        "type": "instantiate",
        "attributes": [
          {
            "key": "_contract_address",
            "value": "wasm1ayj3kc8grpe7axlvdhc0etrhev9knjs0v4ghwc0qn8hda3mdathq75ahs3"
          },
          {
            "key": "code_id",
            "value": "1"
          }
        ]
      },
      {
        "type": "message",
        "attributes": [
          {
            "key": "action",
            "value": "/cosmwasm.wasm.v1.MsgInstantiateContract2"
          },
          {
            "key": "module",
            "value": "wasm"
          },
          {
            "key": "sender",
            "value": "wasm1alz4eapgf0rq3ezf3xge9pz7uhkpdyu2mmrrup"
          }
        ]
      },
      {
        "type": "transfer",
        "attributes": [
          {
            "key": "recipient",
            "value": "wasm1ayj3kc8grpe7axlvdhc0etrhev9knjs0v4ghwc0qn8hda3mdathq75ahs3"
          },
          {
            "key": "sender",
            "value": "wasm1alz4eapgf0rq3ezf3xge9pz7uhkpdyu2mmrrup"
          },
          {
            "key": "amount",
            "value": "100ustake"
          }
        ]
      },
      {
        "type": "wasm",
        "attributes": [
          {
            "key": "_contract_address",
            "value": "wasm1ayj3kc8grpe7axlvdhc0etrhev9knjs0v4ghwc0qn8hda3mdathq75ahs3"
          },
          {
            "key": "Let the",
            "value": "hacking begin"
          }
        ]
      }
    ]
  }
]
```

Reply:
I guess they could also search for type: instantiate?